File: Medium/zigzag_conversion/zigzag_conversion.java
package zigzag_conversion;

import java.util.ArrayList;

class Solution {
    public String convert(String s, int numRows) {
        if(s.length() <= 1 || s.length() > 1000 || numRows <= 1 || numRows > 1000) {
            return s;
        }

        ArrayList<String> collectZigzag = new ArrayList<>();
        int currRow = 0;
        Boolean isStraight = true;

        for(int i = 0; i < s.length(); i++) {
            if(i < numRows) {
                collectZigzag.add("");
            }

            if(currRow == 0) {
                isStraight = true;
            } else if(currRow == numRows - 1) {
                isStraight = false;
            }

            collectZigzag.set(currRow, collectZigzag.get(currRow) + s.charAt(i));

            if(isStraight) {
                currRow++;
            } else {
                currRow--;
            }
        }


        return String.join("", collectZigzag);
    }
}

public class zigzag_conversion {
    public static void main(String[] args) {
        Solution solution = new Solution();

        System.out.println(solution.convert("PAYPALISHIRING", 3));
    }
}

File: Medium/zigzag_conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  if (s.length == 0 || s.length > 1000 || numRows <= 1 || numRows > 1000) {
    return s;
  }

  let collectZigzag = [];
  let isStraight = false;
  let currRow = 0;

  for (let i = 0; i < s.length; i++) {
    if (typeof collectZigzag[currRow] == "undefined") {
      collectZigzag.push("");
    }

    if (currRow == 0) {
      isStraight = true;
    } else if (currRow == numRows - 1) {
      isStraight = false;
    }

    collectZigzag[currRow] += s[i];

    if (isStraight) {
      currRow += 1;
    } else {
      currRow -= 1;
    }
  }
  
  return collectZigzag.join("");
};

console.log(convert("PAYPALISHIRING", 3));
